modules/videogrid-workflowoperation/src/main/java/org/opencastproject/workflow/handler/videogrid/VideoGridWorkflowOperationHandler.java

```diff
@@ -0,0 +1,1007 @@
+    for (URI uri : uris) {
+      TrackImpl track = new TrackImpl();
+      track.setFlavor(targetPresenterFlavor);
+      track.setURI(uri);
+
+      Job inspection = null;
+      try {
+        inspection = inspectionService.enrich(track, true);
+      } catch (MediaInspectionException | MediaPackageException e) {
+        throw new WorkflowOperationException("Inspection service could not enrich track", e);
+      }
+      if (!waitForStatus(inspection).isSuccess()) {
+        throw new WorkflowOperationException(String.format("Failed to add metadata to track."));
+      }
+
+      try {
+        tracks.add((TrackImpl) MediaPackageElementParser.getFromXml(inspection.getPayload()));
+      } catch (MediaPackageException e) {
+        throw new WorkflowOperationException("Could not parse track returned by inspection service", e);
+      }
+    }
+
+    // Concatenate sections
+    Job concatJob = null;
+    try {
+      concatJob = composerService.concat(composerService.getProfile(concatEncodingProfile).getIdentifier(),
+              new Dimension(layoutArea.width,layoutArea.height) , true, tracks.toArray(new Track[tracks.size()]));
+    } catch (EncoderException | MediaPackageException e) {
+      throw new WorkflowOperationException("The concat job failed", e);
+    }
+    if (!waitForStatus(concatJob).isSuccess()) {
+      throw new WorkflowOperationException("The concat job did not complete successfully.");
+    }
+
+    // Add to mediapackage
+    if (concatJob.getPayload().length() > 0) {
+      Track concatTrack;
+      try {
+        concatTrack = (Track) MediaPackageElementParser.getFromXml(concatJob.getPayload());
+      } catch (MediaPackageException e) {
+        throw new WorkflowOperationException("Could not parse track returned by concat service", e);
+      }
+      concatTrack.setFlavor(targetPresenterFlavor);
+      concatTrack.setURI(concatTrack.getURI());
+      for (String tag : targetTags) {
+        concatTrack.addTag(tag);
+      }
+
+      mediaPackage.add(concatTrack);
+    } else {
+      throw new WorkflowOperationException("Concat operation unsuccessful, no payload returned.");
+    }
+
+    try {
+      workspace.cleanup(mediaPackage.getIdentifier());
+    } catch (IOException e) {
+      throw new WorkflowOperationException(e);
+    }
+
+    final WorkflowOperationResult result = createResult(mediaPackage, WorkflowOperationResult.Action.CONTINUE);
+    logger.debug("Video Grid operation completed");
+    return result;
+  }
+
+  /**
+   * Create a ffmpeg command that generate a videos for the given cutting marks
+   * @param layoutArea
+   *          General layout information for the video
+   *          (Originally it was possible to have multiple layout areas)
+   * @param videoEdl
+   *          The edit decision list for the current cut
+   * @return A command line ready ffmpeg command
+   */
+  private List<String> compositeSection(LayoutArea layoutArea, VideoEdlPart videoEdl)
+  {
```

Comment:
It would be nice if you could add some kind of documentation how this algorithm works.

Reply:
Tried to give a general rundown in the method comment